c10/cuda/CUDACachingAllocator.cpp

```diff
@@ -63,34 +64,38 @@ constexpr size_t kLargeBuffer = 20971520;   // "large" allocations may be packed
 constexpr size_t kMinLargeAlloc = 10485760; // allocations between 1 and 10 MiB may use kLargeBuffer
 constexpr size_t kRoundLarge = 2097152;     // round up large allocs to 2 MiB
 
-struct DeviceStats {
-  uint64_t   amount_allocated;      // total amount allocated in bytes
-  uint64_t   max_amount_allocated;  // max total amount allocated in bytes
-  uint64_t   amount_cached;         // total amount in cache in bytes
-  uint64_t   max_amount_cached;     // max total amount in cache in bytes
+typedef std::bitset<static_cast<size_t>(StatType::NUM_TYPES)> StatTypes;
 
-  DeviceStats() :
-      amount_allocated(0), max_amount_allocated(0),
-      amount_cached(0), max_amount_cached(0) { }
+void update_stat(Stat& stat, int64_t amount) {
+  stat.current += amount;
 
-  void increaseAllocated(size_t delta) {
-    amount_allocated += delta;
-    max_amount_allocated = std::max(max_amount_allocated, amount_allocated);
-  }
+  AT_ASSERTM(stat.current >= 0, "Negative tracked stat in CUDA allocator (likely logic error).");
```

Comment:
`TORCH_INTERNAL_ASSERT` here (see https://github.com/pytorch/pytorch/issues/20287 for back story)

Reply:
resolved